File: thememanager/src/main/java/ir/rainyday/thememanager/ThemeManager.java
package ir.rainyday.thememanager;

import android.content.Context;
import android.util.TypedValue;
import android.view.ContextThemeWrapper;


import androidx.annotation.AttrRes;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatDelegate;

public class ThemeManager {

    //region Singleton
    private static volatile ThemeManager instance;
    private static final Object mutex = new Object();

    public static ThemeManager getInstance() {
        ThemeManager result = instance;
        if (result == null) {
            synchronized (mutex) {
                result = instance;
                if (result == null)
                    instance = result = new ThemeManager();
            }
        }
        return result;
    }
    //endregion

    //region Properties
    @Nullable
    private Integer mCurrentTheme = null;

    @Nullable
    public Integer getCurrentTheme() {
        return mCurrentTheme;
    }

    public void setCurrentTheme(@Nullable Integer mCurrentTheme) {
        this.mCurrentTheme = mCurrentTheme;
    }


    @Nullable
    private Integer mNightMode = null;

    @Nullable
    public Integer getNightMode() {
        return mNightMode;
    }

    public void setNightMode(@Nullable Integer mNightMode) {
        this.mNightMode = mNightMode;
    }

    //endregion

    //region Methods
    public TypedValue getTypedValueForAttr(Context context, @AttrRes int attrRes) {
        TypedValue typedValue = new TypedValue();
        context.getTheme().resolveAttribute(attrRes, typedValue, true);
        return typedValue;
    }

    public int getColorPrimary(Context context) {
        return getTypedValueForAttr(context, R.attr.colorPrimary).data;
    }


    public int getColorPrimaryDark(Context context) {
        return getTypedValueForAttr(context, R.attr.colorPrimaryDark).data;
    }

    public int getColorAccent(Context context) {
        return getTypedValueForAttr(context, R.attr.colorAccent).data;
    }

    public void applyTheme(ContextThemeWrapper contextThemeWrapper) {
        if (mCurrentTheme != null)
            contextThemeWrapper.setTheme(mCurrentTheme);

        if (mNightMode != null)
            AppCompatDelegate.setDefaultNightMode(mNightMode);
    }
    //endregion
}

File: app/src/main/java/ir/rainyday/theme_example/ThemeManager.kt
package ir.rainyday.theme_example

import android.content.Context
import android.content.SharedPreferences
import android.preference.PreferenceManager
import android.support.annotation.AttrRes
import android.support.v7.app.AppCompatDelegate
import android.util.TypedValue
import ir.rainyday.theme_example.R


/**
 * Created by mostafa-taghipour on 10/30/17.
 */


class ThemeManager private constructor(var context: Context) {

    private val PREFS_THEME_KEY = "prefs_theme_key"
    private val PREFS_NIGHT_KEY = "prefs_night_key"


    companion object {

        private var instance: ThemeManager? = null

        fun getInstance(context: Context): ThemeManager {
            if (instance == null)  // NOT thread safe!
                instance = ThemeManager(context)

            return instance!!
        }
    }


    private  var prefs:SharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)
    private  var _theme:Int = 0

    init {
        _theme = prefs.getInt(PREFS_THEME_KEY, 0)
    }


    var theme: Int
        get() = _theme
        set(theme) {
            prefs.edit().putInt(PREFS_THEME_KEY, theme).apply()
            _theme = theme
        }

    private var _nightMode: Int = prefs.getInt(PREFS_NIGHT_KEY, AppCompatDelegate.MODE_NIGHT_NO)
    var nightMode: Int
        get() = _nightMode
        set(value) {
            prefs.edit().putInt(PREFS_NIGHT_KEY, value).apply()
            _nightMode = value
        }


    fun getTypedValueForAttr(@AttrRes attrRes: Int): TypedValue {
        val typedValue = TypedValue()
        context.getTheme().resolveAttribute(attrRes, typedValue, true)
        return typedValue
    }

    val colorPrimary: Int
        get() {
            return getTypedValueForAttr(R.attr.colorPrimary).data
        }


    val colorPrimaryDark: Int
        get() {
            return getTypedValueForAttr(R.attr.colorPrimaryDark).data
        }

    val colorAccent: Int
        get() {
            return getTypedValueForAttr(R.attr.colorAccent).data
        }


}

